### PR TITLE
Upgrade login for multi role

### PR DESCRIPTION
**Pull request by sbrunner**

### CODE REVIEW

**Comment by sbrunner** [comment]
Not ready ...

**Comment by sbrunner** [comment]
For me it can be merge as it :-)

File: doc/developer/webservices.rst [hunk 1]
```diff
@@ -279,8 +279,10 @@ Connected JSON result
...
        "username": "<username>",
        "is_password_changed": "True"/"False", # If false the user should change his password
-       "role_name": "<role_name>",
-       "role_id": <role_id>
+       "roles": [{
+           "name": "<role_name>",
+           "id": <role_id>
+       }, ...],
        "functionality": {
            "<functionnality_name>": ["functionnality_value"],
```
Comment (jwkaltz) [inline doc/developer/webservices.rst]: functionality_name and functionality_value
(I know, it's not really part of this PR...)

Comment (sbrunner) [reply doc/developer/webservices.rst]: Fixed :-)

File: geoportal/c2cgeoportal_geoportal/lib/functionality.py [hunk 2]
```diff
@@ -55,12 +56,22 @@ def _get_config_functionality(name, registered, types, request, errors):
...
+        ]
+        return [r for r in result if r is not None]
+
+    result = []
+    for role in user.roles:
+        result.extend([
+            get_typed(name, functionality.value, types, request, errors)
+            for functionality in role.functionalities
+            if functionality.name == name
+        ])
```
Comment (arnaud-morvan) [inline geoportal/c2cgeoportal_geoportal/lib/functionality.py]: You may have many times the same functionnality with same value:
```
result.extend([
        get_typed(name, functionality.value, types, request, errors)
        for functionality in set(
            [f for role in roles for f in role.functionalities if f.name == name]
        )
    ])
```

File: geoportal/tests/functional/test_functionalities.py [hunk 3]
```diff
@@ -127,7 +128,7 @@ def test_functionalities(self):
                 "registered": {},
             },
             "admin_interface": {
-                "available_functionalities": ["__test_a", "__test_s"]
+                "available_functionalities": [{"name": "__test_a"}, {"name": "__test_s"}]
```
Comment (arnaud-morvan) [inline geoportal/tests/functional/test_functionalities.py]: Can it be something else than a name here ?

File: geoportal/c2cgeoportal_geoportal/lib/functionality.py (hunk 2 above)
Comment (arnaud-morvan) [reply geoportal/c2cgeoportal_geoportal/lib/functionality.py]: or maybe better : 
```
result.extend([
        get_typed(name, value, types, request, errors)
        for value in set(
            [f.value for role in roles for f in role.functionalities if f.name == name]
        )
    ])
```

Comment (sbrunner) [reply geoportal/c2cgeoportal_geoportal/lib/functionality.py]: You try to find the more complex solution? For me we should avoid to use `[ ... for ... for ...]`, it's too complex to understand ...

File: geoportal/tests/functional/test_functionalities.py (hunk 3 above)
Comment (sbrunner) [reply geoportal/tests/functional/test_functionalities.py]: Yes we can have: type, regex, error_message, or single, the more important one is single :-)

Comment (arnaud-morvan) [reply geoportal/tests/functional/test_functionalities.py]: Yes, we are in tests, but this is the definitions of functionalities.
My bad.

File: geoportal/c2cgeoportal_geoportal/lib/functionality.py (hunk 2 above)
Comment (arnaud-morvan) [reply geoportal/c2cgeoportal_geoportal/lib/functionality.py]: Ok, so we might create a variable for the values
```
values = set()
for role in roles:
    values = values | set([f.value for f in role.functionalities])
result.extend([
    get_typed(name, value, types, request, errors)
    for value in values
])

Comment (arnaud-morvan) [reply geoportal/c2cgeoportal_geoportal/lib/functionality.py]: This might be more readable:
```
    result.extend([
        get_typed(name, value, types, request, errors)
        for value in set(
            [f.value
             for role in roles
             for f in role.functionalities
             if f.name == name]
        )
    ])
```

Comment (sbrunner) [reply geoportal/c2cgeoportal_geoportal/lib/functionality.py]: Updated with another more readable possibility...

File: geoportal/c2cgeoportal_geoportal/lib/functionality.py [hunk 4]
```diff
@@ -55,13 +56,28 @@ def _get_config_functionality(name, registered, types, request, errors):
...
+        ]
+        return [r for r in values if r is not None]
+
+    else:
+        functionalities = {
+            functionality.value
+            for role in user.roles
+            for functionality in role.functionalities
+            if functionality.name == name
+        }
```
Comment (arnaud-morvan) [inline geoportal/c2cgeoportal_geoportal/lib/functionality.py]: { => [
} => ]
and the first comment was we should use set(functionalities) to have distinct values

Comment (sbrunner) [reply geoportal/c2cgeoportal_geoportal/lib/functionality.py]: {...} directly creates a set ...

**Comment by sbrunner** [comment]
Thanks :-)
